### PR TITLE
dmd: merge into single file

### PR DESCRIPTION
**Pull request by jtbx**
## Description of changes

Merge the `dmd` package into a single file. The bootstrap package has also been merged into its own file.

Also add meta.changelog.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jtbx** [comment]
Result of `nixpkgs-review pr 317845` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/by-name/dm/dmd/package.nix [hunk 1]
```diff
@@ -1,5 +1,216 @@
...
+
+    runHook postInstall
+  '';
+
+  preFixup = ''
+    find $out/bin -type f -exec ${removeReferencesTo}/bin/remove-references-to -t ${dmdBin}/dmd '{}' +
+  '';
+
+  disallowedReferences = [ dmdBootstrap ];
+
```
Comment (Aleksanaa) [inline pkgs/by-name/dm/dmd/package.nix:204]:
```
passthru = { inherit dmdBootstrap };
```

To make it easier to reference the bootstrap package.

Comment (jtbx) [reply pkgs/by-name/dm/dmd/package.nix:204]: Done, thanks!

Comment (jtbx) [reply pkgs/by-name/dm/dmd/package.nix:204]: I've done the same with LDC's bootstrap package too.

**Comment by jtbx** [comment]
@Aleksanaa Could you please take another look at this?

**Comment by CyberShadow** [comment]
Would there be any interest in bootstrapping DMD from source? I've implemented that in Digger.

**Comment by jtbx** [comment]
Sure I'd be up for that. Do you think we could use the existing LDC package to bootstrap it?

**Comment by jtbx** [comment]
> I haven't thought of that. Do you think LDC is simpler to bootstrap from source than DMD?

I thought LDC would be easier due to it not using DMD's D backend. But yes I agree GDC would probably be the best option here.

**Comment by jtbx** [comment]
Superseded by #324982

**Pull request closed without merging**